source/Halibut/Transport/TcpClientManager.cs

```diff
@@ -89,7 +89,9 @@ public void Dispose()
             {
                 if (client.Value?.Any() == true)
                 {
-                    foreach (var tcpClient in client.Value)
+                    var tcpClients = client.Value.ToArray();
```

Comment:
🤔 I'm not sure this will fix the crash.

Looking at the stack trace, it's failing on a `System.Linq.Enumerable.Any`. We have two of those (line 83 and line 90).

But it's also on a `System.Collections.Generic.HashSet`, so that makes it line 90.

However, this fix is after that. Also, whatever is modifying the collection can happen during a `ToArray` just as much as it could happen during `Any`.

From what I can see, this class only modifies this set within locks. 

So I wonder if the fix we are looking for is to wrap the whole dispose methods within a 'lock(activeClients)` block?

Reply:
I suspect that something is interacting with this class on another thread when we decide to dispose of it.

But that in itself is a little concerning.

If we do the lock, and `Dispose` is forced to wait until the other operation finishes, that should be fine.

As long as something else doesn't want to modify this object after we have disposed of it...

Reply:
I should learn to read the error message 😆 !!